Add capability to convert keypoints from COCO to YOLOv5 format

chrisrapson opened the pull request:
In case you're interested in converting keypoints as well as bounding boxes.

Following the format from here:
https://github.com/WongKinYiu/yolov7/issues/1267

I did some manual tests on images with:

- no annotations (outputs a text file with only spaces)
- bounding box annotations but no keypoint annotations (outputs a text file with 5 fields per line, as usual)
- combinations of visible and not-visible keypoints (outputs the usual 5 fields per line, followed by Nx3 values for keypoints)

alexheat commented:
Thank you @chrisrapson I will check it out.

alexheat commented:
@chrisrapson , how is this different than the Segmentation support described here https://github.com/pylabel-project/pylabel/issues/65

chrisrapson commented:
It's definitely similar, but not quite the same. The COCO dataset is the canonical example. They explain the format for keypoints here:
https://cocodataset.org/#format-data

Here's another dataset that has keypoint labels in the VOC format
https://sites.google.com/view/animal-pose/

It looks like it would be challenging to include both segmentation and keypoint data in a YOLO-formatted file, because in a YOLO-formatted file data is interpreted based on its position within a list. Segmentations are an arbitrarily long list of pairs of floats. Keypoints are a list of triplets of floats. The number of keypoints should be the same for all images in a dataset, but won't be the same across datasets. It wouldn't be possible to know when the list of segmentations ended, and the list of keypoints began (or vice versa).

I think it would be simplest to restrict users to convert only one of either keypoints or segmentations to YOLO. I can't think of a use case where somebody would train a network that needs both segmentation _and_ keypoint data. That is, add a flag `keypoints` that has equivalent functionality to your `segmentations` option. Then enforce that maximum one of `segmentations` or `keypoints` can be `True`.

alexheat commented:
Thank you. I had never used key points before but I am getting it. Could you help me with a few more things:
1) Can you recomend a coco dataset that I can use to test it? Should I just get one of the segmentation ones here? https://cocodataset.org/#download
2) Could you add a doc string to the export function that explains this new functionality?

chrisrapson commented:
The keypoints task and the segmentation challenges use the same images. The annotations are saved in `person_keypoints_train2017.json` and `person_keypoints_val2017.json` instead of `instances_train2017.json`. 

One possible place to download them is from huggingface: https://huggingface.co/datasets/merve/coco/tree/main/annotations

Good idea about the doc string. I'll add that and a boolean `keypoints` flag, and then update the PR.

chrisrapson commented:
See the two extra commits. The first adds a docstring and a boolean flag. For the second I added the capability to export keypoints to COCO.

alexheat commented:
I think this is the relevant error message

`DeprecationWarning: Traits should be given as instances, not types (for example, `Int()`, not `Int`). Passing types is deprecated in traitlets 4.1.
    colors = List(Unicode, [
`

alexheat commented:
You can find the instructions to run the tests manually here
https://github.com/pylabel-project/pylabel/tree/dev/tests

alexheat commented:
@chrisrapson I have cherry picked your commits for the yolo output and released it in the latest package, v52. Thank you!

For to Coco export, the issue is in this part of the code:

```
~/Code/scratch/pylabel/pylabel/exporter.py in ExportToCoco(self, output_path, cat_id_index)
    827                 if "ann_keypoints" in df.keys():
    828                     n_keypoints = int(
--> 829                         len(df["ann_keypoints"][i]) / 3
    830                     )  # 3 numbers per keypoint: x,y,visibility
    831                     annotations[0]["num_keypoints"] = n_keypoints
TypeError: object of type 'float' has no len()
```
Is the issue the `[i]` in   `len(df["ann_keypoints"][i])`?

chrisrapson commented:
I think I've fixed it now. I only tested it on my dataset which had keypoint labels for all images. I misunderstood the logic that a dataset (or image) with no keypoint labels wouldn't have `"ann_keypoints"` in its list of `keys`, but of course it has that column filled with `""` which are converted to `np.nan`.

Once I found the test and ran them, it wasn't too hard to implement the if statement properly. I've updated the PR.

alexheat commented:
Thank you @chrisrapson . I Merged it and published it in the latest release .v53. 

It would be awesome to have a sample notebook to demo the functionality for others to add to the library at https://github.com/pylabel-project/samples.

Is that something you would be able to do do (someday)?
